src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -26,80 +26,160 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.BlockingDeque;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.SSLException;
 import javax.net.ssl.TrustManagerFactory;
+import org.apache.commons.lang3.tuple.Pair;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.tikv.common.HostMapping;
 import org.tikv.common.pd.PDUtils;
 
 public class ChannelFactory implements AutoCloseable {
   private static final Logger logger = LoggerFactory.getLogger(ChannelFactory.class);
+  private static final String PUB_KEY_INFRA = "PKIX";
 
+  private int connRecycleTime;
   private final int maxFrameSize;
   private final int keepaliveTime;
   private final int keepaliveTimeout;
   private final int idleTimeout;
-  private final ConcurrentHashMap<String, ManagedChannel> connPool = new ConcurrentHashMap<>();
   private final CertContext certContext;
+
+  @VisibleForTesting
+  public final ConcurrentHashMap<Pair<SslContextBuilder, String>, ManagedChannel> connPool =
+      new ConcurrentHashMap<>();
+
   private final AtomicReference<SslContextBuilder> sslContextBuilder = new AtomicReference<>();
-  private static final String PUB_KEY_INFRA = "PKIX";
 
-  private abstract static class CertContext {
-    protected abstract boolean isModified();
+  private final ScheduledExecutorService recycler = Executors.newScheduledThreadPool(1);
 
-    protected abstract SslContextBuilder createSslContextBuilder();
+  @VisibleForTesting
+  public static class CertWatcher implements AutoCloseable {
+    private static final Logger logger = LoggerFactory.getLogger(CertWatcher.class);
+    private final String path;
+    private final ArrayList<String> targets = new ArrayList<>();
+    private final BlockingDeque<Boolean> notify = new LinkedBlockingDeque<>();
+    private final ExecutorService executorService = Executors.newFixedThreadPool(2);
+
+    public CertWatcher(String path) {
+      if (path == null) {
+        throw new IllegalArgumentException("the base path of certs is missing");
+      }
+      this.path = path;
+      notify.push(true);
+      start();
+    }
 
-    public SslContextBuilder reload() {
-      if (isModified()) {
-        logger.info("reload ssl context");
-        return createSslContextBuilder();
+    private void watchLoop() {
+      try {
+        WatchService watchService = FileSystems.getDefault().newWatchService();
+        Path p = Paths.get(path);
+        p.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);
+
+        WatchKey key;
+        while ((key = watchService.take()) != null) {
```

Comment:
Under which condition would equal null?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/nio/file/WatchService.html#take() should always return notnull seem this null check is redundant.